### ISSUE TITLE
Prometheus Operators incorrectly removes gRPC port from shared prometheus-operated Service resource

### ISSUE THREAD

**Issue by vanugrah** [prometheus-operator/prometheus-operator#6060]
**What happened?**
We deploy hundreds of Prometheus instances leveraging the Prometheus Operator, the majority of which are deployed along with a thanos sidecar. While investigating a separate issue, we made a change to remove `/spec/thanos` from one of the Prometheus instances in the cluster and did not modify the rest (20+) Prometheus instances which did specify `/spec/thanos` in the definition. This change ended up modifying the shared `prometheus-operated` Service resource to remove the gRPC port altogether, which resulted in all of our Prometheus instances being disconnected from our query pipeline. 

**Did you expect to see something different?**
Given that there were several Prometheus instances that still had `/spec/thanos` defined, the gRPC port should have continued to exist on the shared Service resource. 

**How to reproduce it (as minimally and precisely as possible)**:
- Setup multiple Prometheus instances with `/spec/thanos` defined and ensure they are synced and running. 
- Remove `/spec/thanos` from one of the Prometheus instances. 

**Environment**

* Prometheus Operator version:
    `v0.63.0`
* Kubernetes version information:
    `v1.25.7`

**Anything else we need to know?**:

This bug likely happens because the reconciliation of the the Shared Service resource is happening in isolation: 
https://github.com/prometheus-operator/prometheus-operator/blob/ddfea8919b8d489b016bdbd59481a77602cd76eb/pkg/prometheus/server/statefulset.go#L82 

This is fine for Stateful Sets since those resources are unique per instances - however because the Service resource is shared across all Prometheus instances, you need to inspect all of them to determine if there is any Thanos dependency.

**Comment by simonpasquier** [comment]
For now, I'd recommend that you deploy a specific service per Prometheus resource because the default governing service indeed doesn't play well with multiple resources living in the same namespace.
To fix it, we may want to introduce a new `.spec.serviceName` field in the Prometheus(Agent) CRD (as well as Alertmanager & ThanosRuler).

**Comment by simonpasquier** [comment]
_Summarizing discussions from #6067_

I see  a couple of options:

* add a `.spec.serviceName` field to the CRDs:
  * If the field isn't set, the operator defaults to '*-operated' and expects to fully manage the object (same as today).
  * If the service doesn't exist, the operator tries to create it with the same spec as it does for the governing service. The operator doesn't update the service though.
  * If the service exists, the operator does nothing (different from the current approach).
  * Pros
    * Small change to the API.
  * Cons
    * Requires to provision the service in advance.
* embed a full service spec + name in the CRDs:
  * If defined, the operator creates/updates the service object directly.
  * Pros
    * Simpler for users.
  * Cons
    * API surface increase + maintenance burden.

The first option might be enhanced using server-side apply: the operator manages only a few fields and users are free to manipulate unmanaged fields.

**Comment by mviswanathsai** [comment]
Is there any activity on #6067? If the issue is up for grabs, I am interested!

**Comment by mviswanathsai** [comment]
IIUC the user will be allowed to manipulate the unmanaged fields, that are not necessarily "required" by the statefulset. 
1. Could you give an example of what would these fields be? (port, selector?)

Also, something that came to me was:
under the condition that the user changes the port, etc of the Prometheus CR, we would also want to reconcile the managed service accordingly. This would require us to update the service. If we follow that the service will not be updated by the operator (as we do in the first option), in such a condition the svc will not be of much use and would have to be manually changed, but we restrict the changes to Port IIUC. 
Am I mis-understanding this, I see that a lot of what I said is hypothetical without any code/experimentation to back it. But I would like to hear your thoughts before coming to conclusions.

**Comment by mviswanathsai** [comment]
TLDR;
Option 1 might have some edge cases that are not obvious but worth considering. 
Option 2 seems more straight forward but as pointed out by Simon, increases the API surface and comes with it certain burden.

**Comment by simonpasquier** [comment]
After more thoughts, the option 2 is too complex to handle and creates API bloat.

Regarding option 1, I'd simplify even further:
* Add an optional `.spec.serviceName` field to the CRDs.
* If the field isn't set, the operator creates/updates the '*-operated' headless service (clusterIP: "None") and expects to fully manage the object (same as today).
* If `.spec.serviceName` is defined and the service doesn't exist, the operator fails the reconciliation.
* Otherwise, the operator uses its value as the statefulset governing service name.

**Comment by mviswanathsai** [comment]
Should we add to this and ensure that the service with name serviceName does indeed select the given set of Prometheus instances with some set selector. If it does not, fail reconciliation with a relevant error message. 
For example `foo.io/serviceName: foo-bar`

**Comment by simonpasquier** [comment]
@mviswanathsai this could be a nice addition indeed.
